### ISSUE TITLE
Shuttles is slow and UI locks up a lot

### ISSUE THREAD

**Issue by justinanderson** [MIT-Mobile/MIT-Mobile-for-iPhone#425]
Doesn't respond to user interaction while loading data. Happens whenever http fetches are made and pauses the app long enough that it's pretty much unusable.

From a quick look with Instruments, it looks like the slowdown is that all of the CoreData fetching in Shuttles is happening on the main thread. The model for Shuttles should live on a concurrent queue so it can't block anything. If I remember correctly, that's how the model for News works.

On an iPhone 5 running 7.1, it takes more than 30 seconds for a predictions call to finish parsing, during which time the app is frozen. Even on an iPhone 6, the app sometimes pauses for a few seconds when it gets new data.

**Comment by bskinner** [comment]
It's much better than before, but I'm still noticing some pretty significant UI hangups, especially when popping back from one of the detail screens.

**Comment by rosslebeau** [comment]
@bskinner cool, do you want me to take a look on that branch and see if I can find other, non-CD hangups? Or should I wait until you've done all those adjustments and merged into 3.6-dev?

**Comment by bskinner** [comment]
@rosslebeau Since you're more familiar with the behavior of that module than I am at this point, I'd like you to bang on it a bit and see I broke anything else (or introduced some odd behavior).

**Comment by rosslebeau** [comment]
ok I'm taking a look now

**Comment by rosslebeau** [comment]
@bskinner Ok, it looks like a lot of the remaining slowness is due to the Core Data fetching in the MITShuttleMapViewController that is still going on on the main thread, taking up 0.3 seconds or more. The iPhone is a lot faster than iPad because the map is used less, and there less going on when it is used.

I think there are still some things to fix up, but they are tangential to the speed optimization. I'm thinking that using your routes datasource in maps will help a lot, as well as creating a similar thing for vehicles and any other outstanding main thread CD things.

If you have other issues to work on, I'm happy to merge your branch in and use the code as a basis to optimize anything else in shuttles.

**Comment by bskinner** [comment]
@rosslebeau I'll get a PR opened up and get this merged into the main branch if things look fine. After that, go ahead and this code as a basis for fixing up Shuttles as needed.

**Comment by justinanderson** [comment]
@rosslebeau I'm unable to get past the main shuttles screen. Crashes consistently on [MITShuttleStopViewController.m:183](https://github.com/MIT-Mobile/MIT-Mobile-for-iPhone/blob/d89478b253e12a290a1ebf8ef732295b02961a84/Modules/ShuttleTrack/MITShuttleStopViewController.m#L183)

```
2015-01-14 15:44:39.875 MITMobile[70582:5666614] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSArrayM insertObject:atIndex:]: object cannot be nil'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000107538f35 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x00000001071d1bb7 objc_exception_throw + 45
	2   CoreFoundation                      0x000000010740776a -[__NSArrayM insertObject:atIndex:] + 954
	3   MITMobile                           0x000000010209219e -[MITShuttleStopViewController refreshIntersectingRoutes] + 846
	4   MITMobile                           0x0000000102090a81 -[MITShuttleStopViewController initWithStyle:stop:route:] + 257
	5   MITMobile                           0x000000010201d3e3 -[MITShuttleRouteContainerViewController setupStopViewControllers] + 659
	6   MITMobile                           0x000000010201ce81 -[MITShuttleRouteContainerViewController setupChildViewControllers] + 97
	7   MITMobile                           0x000000010201c6a3 -[MITShuttleRouteContainerViewController initWithRoute:stop:] + 499
	8   MITMobile                           0x00000001020ccf83 -[MITShuttleHomeViewController tableView:didSelectRowAtIndexPath:] + 787
	9   UIKit                               0x0000000103cad393 -[UITableView _selectRowAtIndexPath:animated:scrollPosition:notifyDelegate:] + 1293
	10  UIKit                               0x0000000103cad4d4 -[UITableView _userSelectRowAtPendingSelectionIndexPath:] + 219
	11  UIKit                               0x0000000103be8331 _applyBlockToCFArrayCopiedToStack + 314
	12  UIKit                               0x0000000103be81ab _afterCACommitHandler + 516
	13  CoreFoundation                      0x000000010746ddc7 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	14  CoreFoundation                      0x000000010746dd20 __CFRunLoopDoObservers + 368
	15  CoreFoundation                      0x0000000107463b53 __CFRunLoopRun + 1123
	16  CoreFoundation                      0x0000000107463486 CFRunLoopRunSpecific + 470
	17  GraphicsServices                    0x00000001087359f0 GSEventRunModal + 161
	18  UIKit                               0x0000000103bc5420 UIApplicationMain + 1282
	19  MITMobile                           0x000000010201a4fc main + 76
	20  libdyld.dylib                       0x0000000108356145 start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

**Comment by rosslebeau** [comment]
Hm. This didn't occur for me til I wiped the app again. I'm wondering if it's some kind of Core Data context issue...investigating now

**Comment by rosslebeau** [comment]
Nope it's much more mundane than that. These stops have no route because they are orphaned, it seems. They exist in the pre-loaded data but not in the live data. I'll put a fix in ASAP

**Comment by rosslebeau** [comment]
@justinanderson I solved the crash but I actually want to go over this more tomorrow to make sure it's robust. There are still some things about RestKit I'm not sure about.

**Comment by justinanderson** [comment]
@rosslebeau Without knowing much about that code, the Core Data model for Shuttles seems odd to me. All of the relationships are optional and have a nullify delete rule instead of cascade. Also, I noticed there's no inverse relationship for ShuttlePrediction's stop relationship.

And now that I'm able to get down to the other screens on an iPhone, I'm not seeing any annotations on any maps in Shuttles (no route, stops, or vehicles).

Also, that there were orphaned stops between the preload and first load means the preload needs to be updated.

**Comment by rosslebeau** [comment]
@justinanderson I don't know about the optional relationships, but the nullify seems to be a remnant from the old way the model was set up, where stops had multiple parents. Now that I've changed it so that stops have only one parent route, cascade makes more sense for at least that one. As for the ShuttlePrediction's stop relationship, I purposefully removed the inverse because I want the code to always access predictions through a stop's PredictionList. This is not strictly necessary but is more in line with how I want the model to be viewed, so we always have a PredictionList's timestamp in mind, we don't have to set timestamps on each prediction, and enforcing the idea that a Prediction is always owned by a PredictionList makes it easier to avoid issues like orphaned predictions piling up (which seemed to be the cause of the memory leak previously) - without having to make extra fetch requests to have restkit do it, which I figured are good to avoid if possible to help with performance.

I'm happy to take criticism for improvements, as I mostly refactored this code from existing stuff that I wasn't familiar with while trying not to re-write the whole thing, and I'm fairly new to Core Data.

The annotations were not refreshing due to the refactor away from multiple fetchedresultscontrollers. Those are fixed on 3.6-dev now.

**Comment by justinanderson** [comment]
@rosslebeau Lacking an inverse is ok as long as it's intentional.

I wish performance were faster on iPads, but this will be ok for the first release. It's an ok experience on an iPad Air, but anything older than that is pretty laggy.
